include/rocksdb/customizable.h

```diff
@@ -0,0 +1,178 @@
+ * implementations of a class for use by RocksDB (e.g. Table, Cache,
+ * MergeOperator, etc).  The abstract base class is expected to define a method
+ * declaring its type and a factory method for creating one of these, such as:
+ * static const char *Type() { return "Table"; }
+ * static Status CreateFromString(const ConfigOptions& options,
+ *                                const std::string& id,
+ *                                std::shared_ptr<TableFactory>* result);
+ * The "Type" string is expected to be unique (no two base classes are the same
+ * type). This factory is expected, based on the options and id, create and
+ * return the appropriate derived type of the customizable class (e.g.
+ * BlockBasedTableFactory, PlainTableFactory, etc). For extension developers,
+ * helper classes and methods are provided for writing this factory.
+ *
+ * When a Customizable is being created, the "name" property specifies
+ * the name of the instance being created.
+ * For custom objects, their configuration and name can be specified by:
+ * [prop]={name=X;option 1 = value1[; option2=value2...]}
+ *
+ * [prop].name=X
+ * [prop].option1 = value1
+ *
+ * [prop].name=X
+ * X.option1 =value1
+ */
+class Customizable : public Configurable {
+ public:
+  virtual ~Customizable() {}
+  // Returns the name of this class of Customizable
+  virtual const char* Name() const = 0;
+  // Returns an identifier for this Customizable.
+  // This could be its name or something more complex (like its URL/pattern).
+  // Used for pretty printing.
+  virtual std::string GetId() const {
+    std::string id = Name();
+    return id;
+  }
+
+  // This is typically determined by if the input name matches the
+  // name of this object.
+  // This method is typically used in conjunction with GetAs to find the
+  // derived class instance from its base.  For example, if you have an Env
+  // and want the "Default" env, you would IsInstanceOf("Default") to get
+  // the default implementation.  This method should be used when you need a
+  // specific derivative or implementation of a class.
+  //
+  // Intermediary caches (such as SharedCache) may wish to override this method
+  // to check for the intermediary name (SharedCache).  Classes with multiple
+  // potential names (e.g. "PosixEnv", "DefaultEnv") may also wish to override
+  // this method.
+  //
+  // @param name The name of the instance to find.
+  // Returns true if the class is an instance of the input name.
+  virtual bool IsInstanceOf(const std::string& name) const {
+    return name == Name();
+  }
+
+  // Returns the named instance of the Customizable as a T*, or nullptr if not
+  // found. This method uses IsInstanceOf to find the appropriate class instance
+  // and then casts it to the expected return type.
+  template <typename T>
+  const T* GetAs(const std::string& name) const {
+    if (IsInstanceOf(name)) {
+      return static_cast<const T*>(this);
```

Comment:
What happens if I write

    a.GetAs<B>("A")

? Do we not have enough template magic in C++11 to reject this? I expect we could with constexpr functions.

Reply:
This is not using any template magic and is trying to rely on the smarts/accuracy of the programmer.

One way to get around it would be to require all of the T to have a static method (like kName()) and then use that as the "name":
const T* GetAs() const {
  if (IsInstanceOf(T::kName()) {
     ...
}

Note that T::kName() would/should also be returned by the const char *Name() method...

I am not good enough at C++ templates to come up with something better.  I would really like to get rid of all of the assumed casts (most are static_cast) in the code/tests though (the tests do a cast all over for DBImpl).

Reply:
That's basically what I was thinking. For some background / forward-proofing see https://en.wikipedia.org/wiki/Concepts_(C%2B%2B)  I'm using the idea pretty extensively here: https://github.com/pdillinger/rocksdb/blob/ribbon1/util/ribbon_alg.h#L360

Reply:
I will see if I can get it working with a kName() if that is what you think is preferred.  The "concept" qualifiers seem pretty new to C++ and I am not sure how far to take it

Reply:
My recommendation is for informal concepts: informally specified in comments.

Reply:
Seems much better, thanks. (Unless I missed it) more documentation on what's expected of Customizeable classes because of this template would be nice. I might have called it DynamicCast<T> but GetAs<T> is fine with me. :)

Reply:
I will add documentation on the expectations of the Customizable class that have changed since the PR was originated.

I originally named this method CastAs and was asked to change it as part of the PR earlier.  You and @zhichao-cao can fight out which way it should be :)

Reply:
DynamicCast is ok for me. Maybe change to this?

Reply:
Internally it does not do a DynamicCast, so that names seems like a poor naming choice to me.  How about simply "Cast"?

Reply:
The earlier objection to Cast was that "cast" can mean UB if not valid. This is a checked cast returning an appropriate object pointer if valid (ideally) or nullptr if not valid. That is a pointer dynamic cast; we're just doing it without built-in RTTI.

> Internally it does not do a DynamicCast

If we assume there is some logic to the API, why create a function called DynamicCast if you can just dynamic_cast the object in all cases where you can DynamicCast (for the same thing)? (Therefore the function must exist because dynamic_cast might not be available.)

Reply:
Or you could call it CheckedCast.